### PR TITLE
Redeferral checks should also be done on LoopEntryPointInfo's

### PR DESCRIPTION
**Pull request by rajatd**
While deciding whether we can redefer a function body or not, if none of the function entry point infos suggest that the function could be jitted while we redefer the body, we should also do the same checks for loop entry point infos registered with the function body

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/2108)

### CODE REVIEW

**Comment by rajatd** [comment]
@dotnet-bot test this please

File: lib/Runtime/Base/FunctionBody.h [hunk 1]
```diff
@@ -591,6 +591,9 @@ namespace Js
 #endif
     public:
         uint frameHeight;
+#if ENABLE_NATIVE_CODEGEN
+        bool nativeEntryPointProcessed;
```
Comment (pleath) [inline lib/Runtime/Base/FunctionBody.h]: Do you know how we got away with having this only on FunctionEntryPointInfo? Does it never get used on loop entry points?

Comment (rajatd) [inline lib/Runtime/Base/FunctionBody.h]: It was really only used, among other things, to decide whether to expire a function entry point info. LoopEntryPointInfo doesn't have an Expire method.

File: lib/Runtime/Base/FunctionBody.cpp [hunk 2]
```diff
@@ -828,6 +828,22 @@ namespace Js
...
+        if (!isJitCandidate)
+        {
+            // Now check loop body entry points
+            this->MapLoopHeaders([&](uint loopNumber, LoopHeader* header)
+            {
+                header->MapEntryPoints([&](int index, LoopEntryPointInfo* entryPointInfo)
+                {
+                    if ((entryPointInfo->IsCodeGenPending() && isJitModeFunction) || entryPointInfo->IsCodeGenQueued() || entryPointInfo->IsCodeGenRecorded() || (entryPointInfo->IsCodeGenDone() && !entryPointInfo->nativeEntryPointProcessed))
+                    {
+                        isJitCandidate = true;
```
Comment (ThomsonTan) [inline lib/Runtime/Base/FunctionBody.cpp]: >isJitCandidate = true; [](start = 24, length = 22)

break or return from the lambda, and no need to check the remaining loops if any?

Comment (ThomsonTan) [inline lib/Runtime/Base/FunctionBody.cpp]: >isJitCandidate = true; [](start = 24, length = 22)

Does this imply if any loop in a function is JIT candidate, we cannot redefer the whole function?

File: lib/Runtime/Base/FunctionBody.cpp [hunk 3]
```diff
@@ -828,6 +828,22 @@ namespace Js
...
+                    }
+                });
+            });
+        }
+
         if (isJitCandidate)
```
Comment (ThomsonTan) [inline lib/Runtime/Base/FunctionBody.cpp]: >if (isJitCandidate) [](start = 8, length = 19)

Simplify this as "return !isJitCandidate"?

**Comment by pleath** [comment]
LGTM. Thanks.

**Comment by rajatd** [comment]
@dotnet-bot test this please

**Comment by rajatd** [comment]
@dotnet-bot test OSX osx_osx_test please

File: lib/Runtime/Base/FunctionBody.h [hunk 4]
```diff
@@ -591,6 +591,9 @@ namespace Js
...
+#endif
```
Comment (ThomsonTan) [inline lib/Runtime/Base/FunctionBody.h]: >ENABLE_NATIVE_CODEGEN [](start = 4, length = 21)

Why need nativeEntryPointProcessed if ENABLE_NATIVE_CODEGEN is not defined?

Comment (rajatd) [inline lib/Runtime/Base/FunctionBody.h]: That's what I had thought too. But, it has been used in places in the code without checking for ENABLE_NATIVE_CODEGEN. I didn't want to change all those places.

**Pull request merged**